subsys/bluetooth/mesh/pb_gatt_srv.c

```diff
@@ -0,0 +1,348 @@
+#include "pb_gatt_srv.h"
+
+#define CLIENT_BUF_SIZE 66
+
+#if defined(CONFIG_BT_MESH_DEBUG_USE_ID_ADDR)
+#define ADV_OPT_USE_IDENTITY BT_LE_ADV_OPT_USE_IDENTITY
+#else
+#define ADV_OPT_USE_IDENTITY 0
+#endif
+
+#if defined(CONFIG_BT_MESH_PROXY_USE_DEVICE_NAME)
+#define ADV_OPT_USE_NAME BT_LE_ADV_OPT_USE_NAME
+#else
+#define ADV_OPT_USE_NAME 0
+#endif
+
+#define ADV_OPT_PROV                                                           \
+	(BT_LE_ADV_OPT_CONNECTABLE | BT_LE_ADV_OPT_SCANNABLE |                 \
+	 BT_LE_ADV_OPT_ONE_TIME | ADV_OPT_USE_IDENTITY |                       \
+	 BT_LE_ADV_OPT_USE_NAME)
+
+#define ADV_SLOW_INT                                                           \
+	.interval_min = BT_GAP_ADV_SLOW_INT_MIN,                               \
+	.interval_max = BT_GAP_ADV_SLOW_INT_MAX
+
+#define ADV_FAST_INT                                                           \
+	.interval_min = BT_GAP_ADV_FAST_INT_MIN_2,                             \
+	.interval_max = BT_GAP_ADV_FAST_INT_MAX_2
+
+static bool prov_fast_adv;
+
+static int gatt_send(struct bt_conn *conn,
+		     const void *data, uint16_t len,
+		     bt_gatt_complete_func_t end, void *user_data);
+
+static uint8_t __noinit client_buf_data[CLIENT_BUF_SIZE];
+static struct bt_mesh_proxy_role cli = {
+	.cb.send = gatt_send,
+	.buf = {
+		.__buf = client_buf_data,
+		.data  = client_buf_data,
+		.size  = CLIENT_BUF_SIZE,
+		.len   = CLIENT_BUF_SIZE,
+	},
+};
+
+static bool is_register;
+
+static ssize_t gatt_recv(struct bt_conn *conn,
+			 const struct bt_gatt_attr *attr, const void *buf,
+			 uint16_t len, uint16_t offset, uint8_t flags)
+{
+	const uint8_t *data = buf;
+
+	if (cli.conn != conn) {
+		BT_ERR("No PB-GATT Client found");
+		return -ENOTCONN;
+	}
+
+	if (len < 1) {
+		BT_WARN("Too small Proxy PDU");
+		return -EINVAL;
+	}
+
+	if (PDU_TYPE(data) != BT_MESH_PROXY_PROV) {
+		BT_WARN("Proxy PDU type doesn't match GATT service");
+		return -EINVAL;
+	}
+
+	return bt_mesh_proxy_msg_recv(&cli, buf, len);
+}
+
+static void gatt_connected(struct bt_conn *conn, uint8_t err)
+{
+	struct bt_conn_info info;
+
+	bt_conn_get_info(conn, &info);
+	if (info.role != BT_CONN_ROLE_SLAVE ||
+	    !is_register) {
+		return;
+	}
+
+	cli.conn = bt_conn_ref(conn);
+
+	BT_DBG("conn %p err 0x%02x", (void *)conn, err);
+
+	net_buf_simple_reset(&cli.buf);
+}
+
+static void gatt_disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	struct bt_conn_info info;
+
+	bt_conn_get_info(conn, &info);
+	if (info.role != BT_CONN_ROLE_SLAVE ||
+	    !is_register) {
+		return;
+	}
+
+	if (cli.conn != conn) {
+		BT_WARN("No PB-GATT Client found");
+		return;
+	}
+
+	BT_DBG("conn %p reason 0x%02x", (void *)conn, reason);
+
+	bt_mesh_pb_gatt_close(conn);
+
+	if (bt_mesh_is_provisioned()) {
+		(void)bt_mesh_pb_gatt_disable();
+
+		if (IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY) &&
+		    bt_mesh_gatt_proxy_get() !=
+			BT_MESH_GATT_PROXY_NOT_SUPPORTED) {
+			(void)bt_mesh_proxy_gatt_enable();
+		}
+	}
```

Comment:
This could be moved to the proxy server's callback, I think. It'll need to treat the `info.role != BT_CONN_ROLE_SLAVE|| !is_register` part a little different, but it'll eliminate yet another `IS_ENABLED(CONFIG_BT_MESH_GATT_PROXY)` check. It'll also remove the need for ordering the callbacks (I think?)

Reply:
I still think it can be used temporarily, you can modify it at any time if you have a better idea later.

Reply:
Done.